databricks-sdk-java/src/main/java/com/databricks/sdk/core/utils/AzureUtils.java

```
@@ -95,4 +103,87 @@ public static Optional<String> getAzureWorkspaceResourceId(Workspace workspace)
             workspace.getWorkspaceName());
     return Optional.of(resourceId);
   }
+
+  /**
+   * Infers the Azure tenant ID from the Databricks workspace login page.
+   *
+   * @param config The DatabricksConfig instance
+   * @return the discovered tenant ID, or null if discovery fails
+   */
+  public static String inferTenantId(DatabricksConfig config) {
+    if (config.getAzureTenantId() != null) {
+      return config.getAzureTenantId();
+    }
+
+    if (!config.isAzure() || config.getHost() == null) {
+      logger.warn("Cannot infer tenant ID: workspace is not Azure or host is missing");
+      return null;
+    }
+
+    String loginUrl = config.getHost() + AZURE_AUTH_ENDPOINT;
+
+    try {
+      String redirectLocation = getRedirectLocation(config, loginUrl);
+      if (redirectLocation == null) {
+        logger.warn("Failed to get redirect location from Azure auth endpoint: {}", loginUrl);
+        return null;
```

Let's throw an exception instead, the caller should be the one deciding whether these should be logged or not. Same in other places.